ldelf/ta_elf.c

```diff
@@ -262,6 +265,55 @@ static void check_hashtab(struct ta_elf *elf, void *ptr, size_t num_buckets,
 	check_range(elf, "DT_HASH", ptr, sz);
 }
 
+static void check_gnu_hashtab(struct ta_elf *elf, void *ptr)
+{
+	struct gnu_hashtab *h = ptr;
+	uint32_t *bucket = NULL;
+	size_t last_sym = 0;
+	size_t num_words = 4; /* nbuckets, symoffset, bloom_size, bloom_shift */
+	size_t bloom_words = 0;
+	size_t sz = 0;
+	size_t i = 0;
+
+	if (!IS_ALIGNED_WITH_TYPE(ptr, uint32_t))
+		err(TEE_ERROR_BAD_FORMAT, "Bad alignment of DT_GNU_HASH %p",
+		    ptr);
+
+	check_range(elf, "DT_GNU_HASH", h, sizeof(*h));
+
+	if (elf->is_32bit)
+		bloom_words = h->bloom_size;
+	else
+		bloom_words = h->bloom_size * 2;
+
+	if (ADD_OVERFLOW(num_words, h->nbuckets, &num_words) ||
+	    ADD_OVERFLOW(num_words, bloom_words, &num_words) ||
+	    MUL_OVERFLOW(num_words, sizeof(uint32_t), &sz))
+		err(TEE_ERROR_BAD_FORMAT, "DT_GNU_HASH overflow");
+
+	/*
+	 * There is no total number of exported symbols in DT_GNU_HASH,
+	 * so we have to find the highest bucket value manually.
+	 * See http://deroko.phearless.org/dt_gnu_hash.txt.
+	 */
+	bucket = (uint32_t *)h + 4 + bloom_words;
+	check_range(elf, "DT_GNU_HASH", bucket, h->nbuckets * sizeof(uint32_t));
+	last_sym = 0;
+	for (i = 0; i < h->nbuckets; i++)
+		if (bucket[i] > last_sym)
+			last_sym = bucket[i];
+
+	if (MUL_OVERFLOW(last_sym, sizeof(uint32_t), &last_sym) ||
```

Comment:
Is this multiplication needed? IIUC `last_sym` is a word count hence should be added to `num_words`.

Reply:
Sounds right, will fix.